### ISSUE TITLE
Supporting Nested Types?

### ISSUE THREAD

**Issue by lruckman** [PublicApiGenerator/PublicApiGenerator#392]
Has any thought been given to supporting nested types, possibly with a setting to control how deep before bailing?

The [current code](https://github.com/PublicApiGenerator/PublicApiGenerator/blob/11a917b0e9931b553b9df68efc5d7363e8c0a2d7/src/PublicApiGenerator/ApiGenerator.cs#L45) bails immediately after determining the type is nested.

**Comment by sungam3r** [comment]
@lruckman Nope, I don't know the decision behind this. This project was developed a long time ago by @danielmarbach , @stakx and others. Some time ago I became a maintainer of that repo. I think it is better to ask them.

**Comment by danielmarbach** [comment]
Me neither sorry. The only thing I can imagine is that from a public API perspective having deeply nested types seems edge case / awkward territory but I guess there is always some reason why some nesting makes sense 🤷

**Comment by stakx** [comment]
I don't know the history here, either, but I agree with @danielmarbach. Public nested types are quite rare in APIs (about the only scenario I can think of right now are BCL collection enumerator types). I suspect that they simply weren't implemented because supporting them didn't seem crucial for most use cases.

(If support for them gets added in this project, I think it would be strange to have a configurable nesting depth limit. For one thing, nested types are quite rare, so more deeply nested types would probably be extremely rare... not much point in making too big of an effort for such edge cases. And more importantly, the main criterion should be whether the type is visible/accessible or not, no matter its nesting depth.)

**Comment by sungam3r** [comment]
Thanks. I will try to experiment with allowing nested types in generated API.

**Comment by lruckman** [comment]
The nesting limit would be more of a safety net to avoid infinitely recursing through a type that has children of the same type. Or maybe just an option to error or allow something like what  Newtonsoft has `ReferenceLoopHandling`. 

In terms of does it work or not I’ve been testing removing the check (`!typeDefinition.IsNested`) via reflection and it seems to work great so far.

**Comment by stakx** [comment]
> The nesting limit would be more of a safety net to avoid infinitely recursing through a type that has children of the same type.

I don't think this is possible: A type cannot be nested within itself, neither directly nor indirectly.

Are you perhaps confusing recursive types (via type references as in e.g. `interface Node { Node Next { get; } }`) with nested type declarations (`class Outer { class Inner { } }`)? The latter cannot be recursive.

**Comment by lruckman** [comment]
Yeah sorry, early morning.

**Comment by sungam3r** [comment]
Closing this one as it seems there is nothing to do here.
